base/applications/cmdutils/tasklist/tasklist.c

```diff
@@ -0,0 +1,370 @@
+/*
+ * PROJECT:     ReactOS Tasklist Command
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Displays a list of currently running processes on the computer.
+ * COPYRIGHT:   Copyright 2020 He Yang (1160386205@qq.com)
+ */
+
+#include "tasklist.h"
+
+
+// the strings in opList are the command-line options.
+// should always correspond with the defines below, in sequence (except OP_PARAM_INVALID)
+static PCWSTR opList[] = { L"?", L"nh" };
+
+#define OP_PARAM_INVALID    -1
+#define OP_PARAM_HELP       0
+#define OP_PARAM_NOHEADER    1
+
+// the max string length PrintResString can handle
+#define RES_STR_MAXLEN 64
+
+VOID PrintSplitLine(INT Length)
+{
+    for (INT i = 0; i < Length; i++)
+    {
+        ConPrintf(StdOut, L"=");
+    }
+}
+
+// Print spaces
+BOOL PrintSpace(INT SpaceNum)
+{
+    for (; SpaceNum; SpaceNum--)
+    {
+        ConPrintf(StdOut, L" ");
+    }
+    return TRUE;
+}
+
+// Print a string.
+// if bAlignLeft == TRUE then aligned to left, otherwise aligned to right
+// MaxWidth is the width for printing.
+INT PrintString(LPCWSTR String, INT MaxWidth, BOOL bAlignLeft)
+{
+    return ConPrintf(StdOut, bAlignLeft ? L"%-*.*ls" : L"%*.*ls", MaxWidth, MaxWidth, String);
+}
+
+// Print a string from resource
+// if bAlignLeft == TRUE then aligned to left, otherwise aligned to right
+// MaxWidth is the width for printing.
+// The string will be truncated if it's longer than RES_STR_MAXLEN
+INT PrintResString(HINSTANCE hInstance, UINT uid, INT MaxWidth, BOOL bAlignLeft)
+{
+    WCHAR StringBuffer[RES_STR_MAXLEN];
+    LoadStringW(hInstance, uid, StringBuffer, _countof(StringBuffer));
+    return PrintString(StringBuffer, MaxWidth, bAlignLeft);
+}
+
+// Print a number, aligned to right.
+// MaxWidth is the width for printing.
+// return FALSE if number length is longer than MaxWidth
+BOOL PrintNum(LONGLONG Number, INT MaxWidth)
+{
+    // calculate the length needed to print.
+    INT PrintLength = 0;
+    LONGLONG Tmp = Number;
+
+    if (Tmp < 0)
+    {
+        Tmp = -Tmp;
+        PrintLength++;
+    }
+
+    do
+    {
+        Tmp /= 10;
+        PrintLength += 1;
+    } while (Tmp);
+
+    if (PrintLength > MaxWidth)
+    {
+        return FALSE;
+    }
+    ConPrintf(StdOut, L"%*lld", MaxWidth, Number);
+
+    return TRUE;
+}
+
+// Print memory size using appropriate unit, with comma-separated number, aligned to right.
+// MaxWidth is the width for printing.
+// StartingUnit is the minimum memory unit used, 0 for Byte, 1 for KB, 2 for MB ...
+// return FALSE when failed to format.
+BOOL PrintMemory(SIZE_T MemorySizeByte, INT MaxWidth, INT StartingUnit)
+{
+    WCHAR *MemoryUnit[] = {L"B", L"K", L"M", L"G", L"T", L"P"};
+
+    MaxWidth -= 2; // a space and a unit will occupy 2 length.
+    if (MaxWidth <= 0)
+    {
+        return FALSE;
+    }
+
+    SIZE_T MemorySize = MemorySizeByte;
+
+    for (INT i = 0; i < _countof(MemoryUnit); i++)
+    {
+        if (i >= StartingUnit)
```

Comment:
then I also have to do ```MemorySize >>= 10;``` before continuing, I don't want to write this twice 🤣 (should I write this in the for loop instead?)
by the way, I'm wondering what does the unit ```K``` in tasklist refers to?
AFAIN it can refer to ```KB``` or ```KiB```, and that's two different units.....

Reply:
The notation of these prefixes is also language/locale-dependent.

Reply:
prefixes ?

Reply:
Errm I meant these suffixes (the K, M, G, etc...)

Reply:
the unit is locale-dependent?! 
Oh, I don't know this before...
should I move them into the resources file or what?

Reply:
`<shlwapi.h>` has `StrFormatByteSizeW` function.
https://docs.microsoft.com/ja-jp/windows/win32/api/shlwapi/nf-shlwapi-strformatbytesizew

Reply:
Be careful however to not have core command-line utilities depend on optional GUI shell features though.

Reply:
StrFormatByteSizeW has no GUI. Should we use StrFormatByteSizeW? Otherwise what we use?

Reply:
ah, I've recently checked the resources that the  ```tasklist``` on windows have included, and I have found this:
```
    253, "%s K"
```
guess this is what tasklist used to format the memory usage string? BTW I didn't found any other strings like M,G,T.
possibly it's supposed to only display memory in KB/KiB?

Reply:
Yes do that. I would suggest (because some people could invoke "compatibility" with the Windows tasklist and tools that could parse its output) to default to only displaying the numbers in KB (multiple of 1024), and only if you specify a ROS-specific switch on the command-line, would you format the output in KB/MB/GB etc... depending on the size.

Reply:
> Yes do that. I would suggest (because some people could invoke "compatibility" with the Windows tasklist and tools that could parse its output) to default to only displaying the numbers in KB (multiple of 1024), and only if you specify a ROS-specific switch on the command-line, would you format the output in KB/MB/GB etc... depending on the size.

Ok, I will refactor it tonight.